### PR TITLE
add actions

### PR DESCRIPTION
**Pull request by tobowers**
This adds the ability to use "actions" in text (and this also includes #22 so can replace that PR).

So far I have only tried RAMBLE and it appears to be working really well.

### CODE REVIEW

File: core/examples/example.ts [hunk 1]
```diff
@@ -2,7 +2,9 @@ import * as readline from "readline";
 import { Blueprints, Soul } from "../src";
 
 const blueprint = Blueprints.SAMANTHA;
-const soul = new Soul(Blueprints.SAMANTHA);
+const soul = new Soul(Blueprints.SAMANTHA, {
+  additionalContext: `Samantha is texting someone new at MeetSamantha.ai. The time and date is: ${new Date().toLocaleString()}}`,
```
Comment (kafischer) [inline core/examples/example.ts:6]: Should this be a conversation specific context?

File: core/src/TEMPLATES.ts [hunk 2]
```diff
@@ -29,7 +44,8 @@ Think {
    <MESSAGE>...</MESSAGE>
    <RETROSPECTION>In retrospect, ... </RETROSPECTION>
    <NEXT_PLAN>My next plan is ... </NEXT_PLAN>
-   <WANTS_TO_RAMBLE>(YES or NO)</WANTS_TO_RAMBLE>
+   <ACTION>One of ${actionNames}</ACTION>
+   <ACTION_INPUT>Any input to the action</ACTION_INPUT>
```
Comment (kafischer) [inline core/src/TEMPLATES.ts:48]: Nice! I even imagine that it generates input notes to its future self for ramble?

File: core/src/conversationProcessor.ts [hunk 3]
```diff
@@ -59,6 +68,30 @@ export class ConversationProcessor extends EventEmitter {
...
+        }
+        this.generatedThoughts.push(
+          new Thought({
+            role: "assistant",
+            entity: this.blueprint.name,
+            action: "RAMBLE",
+            content: "I want to ramble before they respond",
+          }),
+        );
+        this.continueThinking();
```
Comment (kafischer) [inline core/src/conversationProcessor.ts:89]: would this be `conversation.genereatedThoughts ...`, `conversation.continueThinking()`, and `soul.actions`

with the signature as in the action interface for clarity?

**Review by kafischer — CHANGES_REQUESTED** [review]
AMAZING very cool addition

File: core/src/TEMPLATES.ts (hunk 2 above)
Comment (tobowers) [reply core/src/TEMPLATES.ts:48]: In my current ramble method I ask it to say "yes" or "no" but yeah notes are smart. One thing that I want to figure out on the next version of this is putting the action and the ACTION_INPUT together.

File: core/src/conversationProcessor.ts (hunk 3 above)
Comment (tobowers) [reply core/src/conversationProcessor.ts:89]: I'm not sure exactly what you're saying here, but I think you're saying add the other two inputs next to "input"?

File: core/examples/example.ts (hunk 1 above)
Comment (tobowers) [reply core/examples/example.ts:6]: made this true, and showed the power of dynamic contexts in the example.

File: core/src/conversationProcessor.ts (hunk 3 above)
Comment (tobowers) [reply core/src/conversationProcessor.ts]: I changed it to show the full signature - let me know if that's not what you meant.

File: core/src/conversationProcessor.ts [hunk 4]
```diff
@@ -59,6 +68,30 @@ export class ConversationProcessor extends EventEmitter {
...
+        }
+        this.generatedThoughts.push(
+          new Thought({
+            role: "assistant",
+            entity: this.blueprint.name,
+            action: "RAMBLE",
+            content: "I want to ramble before they respond",
+          })
+        );
+        this.continueThinking();
```
Comment (kafischer) [inline core/src/conversationProcessor.ts:89]: Am I misunderstanding or wouldn't this be more clear?

```suggestion
        conversation.generatedThoughts.push(
          new Thought({
            role: "assistant",
            entity: this.blueprint.name,
            action: "RAMBLE",
            content: "I want to ramble before they respond",
          })
        );
        conversation.continueThinking();
```

Comment (tobowers) [reply core/src/conversationProcessor.ts]: ah good idea. made that change.

**Review by kafischer — APPROVED** [review]
Wohoo!

**Pull request merged**